src/sources.rs

```diff
@@ -0,0 +1,346 @@
+    let parser = |v: CoinGeckoResponse| Some(v.concordium.eur);
+    request_exchange_rate_aux(client.get(COINGECKO_URL), parser, config::COINGECKO_LABEL).await
+}
+
+async fn request_exchange_rate_livecoinwatch(
+    client: reqwest::Client,
+    api_key: String,
+) -> Option<f64> {
+    let params = json!({"currency":"EUR","code":"CCD","meta":false});
+    let request = client.post(LIVECOINWATCH_URL).json(&params).header("x-api-key", api_key);
+    let parser = |v: LiveCoinWatchResponse| Some(v.rate);
+    request_exchange_rate_aux(request, parser, config::LIVECOINWATCH_LABEL).await
+}
+
+async fn request_exchange_rate_coinmarketcap(
+    client: reqwest::Client,
+    api_key: String,
+) -> Option<f64> {
+    let request = client.get(COINMARKETCAP_URL).header("X-CMC_PRO_API_KEY", api_key);
+    let parser = |v: CoinMarketCapResponse| Some(v.data.ccd[0].quote.eur.price);
+    request_exchange_rate_aux(request, parser, config::COINMARKETCAP_LABEL).await
+}
+
+/**
+ * Pulls the exchange rate using the provided client from the given source.
+ */
+async fn request_matcher(client: reqwest::Client, source: Source) -> Option<f64> {
+    match source {
+        Source::Bitfinex => request_exchange_rate_bitfinex(client).await,
+        Source::LiveCoinWatch(api_key) => {
+            request_exchange_rate_livecoinwatch(client, api_key).await
+        }
+        Source::CoinMarketCap(api_key) => {
+            request_exchange_rate_coinmarketcap(client, api_key).await
+        }
+        Source::CoinGecko => request_exchange_rate_coingecko(client).await,
+        Source::Test(url, id) => {
+            request_exchange_rate_aux(client.get(url.clone()), |v: Vec<f64>| Some(v[0]), &id).await
+        }
+    }
+}
+
+/**
+ * Function that continously pulls the exchange rate, from the source
+ * specified, and updates the given rates_history_mutex. Ensures that old
+ * rates are discarded, when the queue exceeds max size.
+ */
+pub async fn pull_exchange_rate(
+    stats: prometheus::Stats,
+    source: Source,
+    rate_history_mutex: Arc<Mutex<VecDeque<BigRational>>>,
+    pull_interval: u32,
+    max_rates_saved: usize,
+    mut database_conn: Option<mysql::PooledConn>,
+) -> anyhow::Result<()> {
+    let client = reqwest::Client::new();
+    let source_label = match source {
+        Source::Bitfinex => config::BITFINEX_LABEL,
+        Source::LiveCoinWatch(_) => config::LIVECOINWATCH_LABEL,
+        Source::CoinMarketCap(_) => config::COINMARKETCAP_LABEL,
+        Source::CoinGecko => config::COINGECKO_LABEL,
+        Source::Test(_, ref id) => &id,
+    };
+
+    let mut interval = interval(Duration::from_secs(pull_interval.into()));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    loop {
+        interval.tick().await;
+        log::debug!("{}: Polling for exchange rate", source_label);
+
+        let raw_rate = match request_with_backoff(
+            || request_matcher(client.clone(), source.clone()),
```

Comment:
Is the cloning of source necessary, could we not do without?

This cloning means copying API keys all over the memory without properly clearing them on drop, which is not the best of practices. If one cannot get rid of the clone then wrap the strings into an Arc, so data is not copied, only the pointer.